src/metabase/driver/sql/query_processor.clj

```diff
@@ -1021,41 +1154,75 @@
+  query), or some sort of expression."
+  ([driver honeysql-form]
+   (format-honeysql (honey-sql-version driver) (quote-style driver) honeysql-form))
+
+  ([version dialect honeysql-form]
+   (try
+     (let [f (case (long version)
+               1 #_{:clj-kondo/ignore [:deprecated-var]} sql.qp.deprecated/format-honeysql-1
+               2 format-honeysql-2)]
+       (f dialect honeysql-form))
+     (catch Throwable e
+       (try
+         (log/error e
+                    (u/format-color 'red
+                                    (str (deferred-tru "Invalid HoneySQL form: {0}" (ex-message e))
+                                         "\n"
+                                         (u/pprint-to-str honeysql-form))))
+         (finally
+           (throw (ex-info (tru "Error compiling HoneySQL form: {0}" (ex-message e))
+                           {:dialect           dialect
+                            :form              honeysql-form
+                            :type              qp.error-type/driver
+                            :honey-sql-version version}
+                           e))))))))
+
+(defn- default-select [driver {[from] :from, :as _honeysql-form}]
+  (let [table-identifier (if (sequential? from)
+                           ;; Grab the alias part.
+                           ;;
+                           ;; Honey SQL 1 = [expr alias]
+                           ;; Honey SQL 2 = [expr [alias]]
+                           (case hx/*honey-sql-version*
+                             1 (second from)
+                             2 (first (second from)))
+                           from)
+        [raw-identifier] (format-honeysql driver table-identifier)
+        expr             (if (seq raw-identifier)
+                           (hx/raw (format "%s.*" raw-identifier))
+                           :*)]
+    [(maybe-wrap-unaliased-expr expr)]))
 
 (defn- add-default-select
   "Add `SELECT *` to `honeysql-form` if no `:select` clause is present."
-  [driver {:keys [select], [from] :from, :as honeysql-form}]
+  [driver {:keys [select], :as honeysql-form}]
   ;; TODO - this is hacky -- we should ideally never need to add `SELECT *`, because we should know what fields to
   ;; expect from the source query, and middleware should be handling that for us
   (cond-> honeysql-form
-    (empty? select) (assoc :select (let [table-identifier (if (sequential? from)
-                                                            (second from)
-                                                            from)
-                                         [raw-identifier] (format-honeysql driver table-identifier)]
-                                     (if (seq raw-identifier)
-                                       [(hx/raw (format "%s.*" raw-identifier))]
-                                       [:*])))))
+    (empty? select) (assoc :select (default-select driver honeysql-form))))
```

Comment:
This is one reason why `format-honeysql` takes a snippet, I wonder if it would be possible to drop this code and the snippet handling, with some well placed asserts.

Reply:
Yeah, that might make sense. In some cases I think we actually do hit this path for full queries tho. We can try taking it out and seeing if anything breaks